akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala

```diff
@@ -159,9 +175,12 @@ object ShardCoordinator {
           case (_, v) ⇒ v.filterNot(s ⇒ rebalanceInProgress(s))
         }.maxBy(_.size)
         val difference = mostShards.size - leastShards.size
-        if (difference >= rebalanceThreshold)
-          Future.successful(mostShards.take(math.min(difference, maxSimultaneousRebalance - rebalanceInProgress.size)).toSet)
-        else
+        if (difference > rebalanceThreshold) {
+          val n = math.min(
+            math.min(difference - rebalanceThreshold, rebalanceThreshold),
```

Comment:
Previously the problem was that it was too aggressive. For example if there was 5 shards per region and a new region was added the difference would be 5, meaning that 5 shards would be stopped, and then later 3 or 2 of them would be allocated to the same region as they were originally located.

This will do it more step by step, e.g. if threshold is 2 in the above example it would do the right thing by rebalancing 2 from one of the nodes. If threshold is 1 it would take 1 and then later 1 more.

The algorithm isn't always perfect, and can sometimes still rebalance too much than is theoretically optimal but as mentioned that is out of scope for this ticket and probably not important in practice.

I took a rather test driven approach to this problem and added many scenarios and thereby came up with this algorithm that seemed good for most of them.

Reply:
Ok, thx for explaining.